core/src/main/java/io/questdb/cairo/ColumnVersionPurgeJob.java

```diff
@@ -0,0 +1,349 @@
+                        "partition_name_txn long," + // 10
+                        "completed timestamp" + // 11
+                        ") timestamp(ts) partition by MONTH",
+                sqlExecutionContext
+        );
+        this.writer = engine.getWriter(AllowAllCairoSecurityContext.INSTANCE, getLogTableName(), "QuestDB system");
+        this.cleanupExecution = new ColumnVersionPurgeExecution(configuration, this.writer, "completed");
+    }
+
+    @Override
+    public void close() {
+        this.writer = Misc.free(this.writer);
+        this.sqlCompiler = Misc.free(sqlCompiler);
+        this.sqlExecutionContext = Misc.free(sqlExecutionContext);
+        this.cleanupExecution = Misc.free(cleanupExecution);
+    }
+
+    public String getLogTableName() {
+        return tableName;
+    }
+
+    private static int compareHouseKeepingTasks(ColumnVersionPurgeTaskRun task1, ColumnVersionPurgeTaskRun task2) {
+        return Long.compare(task1.nextRunTimestamp, task2.nextRunTimestamp);
+    }
+
+    private void calculateNextTimestamp(ColumnVersionPurgeTaskRun task, long currentTime) {
+        task.waitToRun = Math.min(maxWaitCapMicro, (long) (task.waitToRun * exponentialWaitMultiplier));
+        task.nextRunTimestamp = currentTime + task.waitToRun;
+    }
+
+    private boolean cleanup() {
+        boolean useful = false;
+        final long now = clock.getTicks() + 1;
+        while (houseKeepingRunQueue.size() > 0) {
+            ColumnVersionPurgeTaskRun next = houseKeepingRunQueue.peek();
+            if (next.nextRunTimestamp < now) {
+                houseKeepingRunQueue.poll();
+                useful = true;
+                if (!cleanupExecution.tryCleanup(next)) {
+                    // Re-queue
+                    calculateNextTimestamp(next, now);
+                    houseKeepingRunQueue.add(next);
+                }
+            } else {
+                // All reruns are in the future.
+                return useful;
+            }
+        }
+        return useful;
+    }
+
+    private void commit() {
+        try {
+            writer.commit();
+        } catch (Throwable th) {
+            LOG.error().$("error saving to column version house keeping log, cannot commit [table=").$(tableName).$(", error=").$(th).I$();
+            writer.rollback();
+            writer = Misc.free(writer);
+            throw th;
+        }
+    }
+
+    // Process incoming queue and put it on priority queue with next timestamp to rerun
+    private boolean processInQueue() {
+        boolean any = false;
+        long microTime = clock.getTicks();
+        while (true) {
+            long cursor = inSubSequence.next();
+            // -2 = there was a contest for queue index and this thread has lost
+            if (cursor < -1) {
+                continue;
+            }
+
+            // -1 = queue is empty. All done.
+            if (cursor < 0) {
+                if (any) {
+                    commit();
```

Comment:
moved commit() out of the loop and renamed ```any``` to ```useful```.
did not replace it with ```writer.inTransaction()``` because ```writer``` could be null if it is freed because of an error.